tool/tsh/db.go

```diff
@@ -91,73 +120,175 @@ func onListDatabases(cf *CLIConf) error {
-	l[i], l[j] = l[j], l[i]
+func (l *databaseListings) Swap(i, j int) {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	l.listings[i], l.listings[j] = l.listings[j], l.listings[i]
 }
 
 func listDatabasesAllClusters(cf *CLIConf) error {
-	var dbListings databaseListings
-	err := forEachProfile(cf, func(tc *client.TeleportClient, profile *client.ProfileStatus) error {
-		result, err := tc.ListDatabasesAllClusters(cf.Context, nil /* custom filter */)
-		if err != nil {
-			return trace.Wrap(err)
-		}
+	dbListings := &databaseListings{}
+	errors := syncErrors{}
+	waitAll := sync.WaitGroup{}
 
-		proxy, err := tc.ConnectToProxy(cf.Context)
-		if err != nil {
-			return trace.Wrap(err)
-		}
-		var errors []error
-		for clusterName, databases := range result {
-			cluster, err := proxy.ConnectToCluster(cf.Context, clusterName)
-			if err != nil {
-				errors = append(errors, err)
-				continue
-			}
-
-			roleSet, err := services.FetchAllClusterRoles(cf.Context, cluster, profile.Roles, profile.Traits)
-			if err != nil {
-				log.Debugf("Failed to fetch user roles: %v.", err)
+	err := forEachProfile(cf, func(tc *client.TeleportClient, profile *client.ProfileStatus) error {
+		waitAll.Add(1)
+		go func() {
+			defer waitAll.Done()
+
+			// Run ListDatabasesAllClusters and ConnectToProxy in parallel and wait.
+			errorsPerProfile := syncErrors{}
+			waitPerProfile := sync.WaitGroup{}
+			waitPerProfile.Add(2)
+			var databasesByCluster map[string][]types.Database
+			var proxyClient *client.ProxyClient
+
+			go func() {
+				defer waitPerProfile.Done()
+
+				result, err := tc.ListDatabasesAllClusters(cf.Context, nil /* custom filter */)
+				if err != nil {
+					errorsPerProfile.Add(trace.Wrap(err))
+					return
+				}
+
+				databasesByCluster = result
+			}()
+
+			go func() {
+				defer waitPerProfile.Done()
+
+				proxy, err := tc.ConnectToProxy(cf.Context)
+				if err != nil {
+					errorsPerProfile.Add(trace.Wrap(err))
+					return
+				}
+
+				proxyClient = proxy
+			}()
+
+			waitPerProfile.Wait()
+			if errorsPerProfile.Len() > 0 {
+				errors.Add(errorsPerProfile.ToSlice()...)
+				return
 			}
 
-			for _, database := range databases {
-				dbListings = append(dbListings, databaseListing{
-					Proxy:    profile.ProxyURL.Host,
-					Cluster:  clusterName,
-					roleSet:  roleSet,
-					Database: database,
-				})
+			// Add fetch roles for each cluster to waitAll.
+			for clusterName, databases := range databasesByCluster {
+				waitAll.Add(1)
+				go func() {
+					defer waitAll.Done()
+
+					cluster, err := proxyClient.ConnectToCluster(cf.Context, clusterName)
+					if err != nil {
+						errors.Add(trace.Wrap(err))
+						return
+					}
+
+					roleSet, err := services.FetchAllClusterRoles(cf.Context, cluster, profile.Roles, profile.Traits)
+					if err != nil {
+						log.Debugf("Failed to fetch user roles: %v.", err)
+					}
+
+					for _, database := range databases {
+						dbListings.Add(databaseListing{
```

Comment:
This has far too much synchronization - we can just give each goroutine a slice of `databaseListing` to fill, can't we? Locking a rwlock (or even a mutex) on every swap in a sort would be quite slow.